File: specs/examples/typescript/auth-controller.ts
/**
 * Authentication Controller - TypeScript Example
 * 
 * This is an EXAMPLE implementation based on the User Authentication feature spec.
 * Customize this for your framework (Express, Fastify, etc.).
 * 
 * Spec Reference: specs/features/example-feature.md
 * API Contract: specs/api/openapi.yaml
 */

import { Request, Response } from 'express';
import { AuthService, LoginRequest } from './auth-login-service';

export class AuthController {
  constructor(private authService: AuthService) {}

  /**
   * POST /auth/login
   * 
   * Implements: specs/features/example-feature.md
   * API Contract: specs/api/openapi.yaml#/paths/~1auth~1login
   * 
   * Request Example: specs/examples/requests/login-request.json
   * Response Example: specs/examples/responses/login-success.json
   */
  async login(req: Request, res: Response): Promise<void> {
    try {
      const request: LoginRequest = {
        username: req.body.username,
        password: req.body.password
      };

      // Get client IP for rate limiting
      const ipAddress = req.ip || req.socket.remoteAddress || 'unknown';

      const result = await this.authService.login(request, ipAddress);

      if (result.success) {
        res.status(200).json(result);
      } else {
        // Map error codes to HTTP status codes
        const statusCode = this.mapErrorToStatusCode(result.error?.code || 'UNKNOWN');
        res.status(statusCode).json(result);
      }
    } catch (error) {
      res.status(500).json({
        success: false,
        data: null,
        error: {
          code: 'INTERNAL_ERROR',
          message: 'An error occurred processing your request'
        }
      });
    }
  }

  private mapErrorToStatusCode(errorCode: string): number {
    const errorMap: Record<string, number> = {
      'VALIDATION_ERROR': 400,
      'INVALID_CREDENTIALS': 401,
      'ACCOUNT_LOCKED': 423,
      'RATE_LIMIT_EXCEEDED': 429,
      'UNAUTHORIZED': 401,
      'FORBIDDEN': 403,
      'NOT_FOUND': 404
    };

    return errorMap[errorCode] || 500;
  }
}

File: specs/examples/python/auth_controller.py
"""
Authentication Controller - Python Example (FastAPI)

This is an EXAMPLE implementation based on the User Authentication feature spec.
Customize this for your framework (FastAPI, Flask, Django, etc.).

Spec Reference: specs/features/example-feature.md
API Contract: specs/api/openapi.yaml
"""

from fastapi import APIRouter, Request, HTTPException, status
from pydantic import BaseModel
from .auth_login_service import AuthService, LoginRequest, LoginResponse

router = APIRouter(prefix="/auth", tags=["Authentication"])


@router.post(
    "/login",
    response_model=LoginResponse,
    summary="Login with credentials",
    description="Authenticate user with username/password",
    responses={
        200: {
            "description": "Login successful",
            "content": {
                "application/json": {
                    "example": {
                        "success": True,
                        "data": {
                            "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...",
                            "expiresAt": "2025-01-17T12:00:00Z",
                            "user": {
                                "id": "user_123abc",
                                "username": "alice",
                                "email": "alice@example.com"
                            }
                        },
                        "error": None
                    }
                }
            }
        },
        401: {
            "description": "Invalid credentials",
            "content": {
                "application/json": {
                    "example": {
                        "success": False,
                        "data": None,
                        "error": {
                            "code": "INVALID_CREDENTIALS",
                            "message": "Invalid username or password"
                        }
                    }
                }
            }
        },
        429: {
            "description": "Too many requests",
            "content": {
                "application/json": {
                    "example": {
                        "success": False,
                        "data": None,
                        "error": {
                            "code": "RATE_LIMIT_EXCEEDED",
                            "message": "Too many login attempts"
                        }
                    }
                }
            }
        }
    }
)
async def login(
    request: LoginRequest,
    http_request: Request
) -> LoginResponse:
    """
    POST /auth/login
    
    Implements: specs/features/example-feature.md
    API Contract: specs/api/openapi.yaml#/paths/~1auth~1login
    
    Request Example: specs/examples/requests/login-request.json
    Response Example: specs/examples/responses/login-success.json
    """
    # Get client IP for rate limiting
    ip_address = http_request.client.host if http_request.client else "unknown"
    
    # Initialize service (in production, use dependency injection)
    auth_service = AuthService(
        user_repository=None,  # Inject your repository
        token_service=None,     # Inject your token service
        rate_limiter=None       # Inject your rate limiter
    )
    
    result = await auth_service.login(request, ip_address)
    
    if not result.success:
        # Map error codes to HTTP status codes
        status_code = _map_error_to_status_code(result.error.get("code") if result.error else "UNKNOWN")
        raise HTTPException(status_code=status_code, detail=result.error)
    
    return result


def _map_error_to_status_code(error_code: str) -> int:
    """Map error codes to HTTP status codes"""
    error_map = {
        "VALIDATION_ERROR": status.HTTP_400_BAD_REQUEST,
        "INVALID_CREDENTIALS": status.HTTP_401_UNAUTHORIZED,
        "ACCOUNT_LOCKED": status.HTTP_423_LOCKED,
        "RATE_LIMIT_EXCEEDED": status.HTTP_429_TOO_MANY_REQUESTS,
        "UNAUTHORIZED": status.HTTP_401_UNAUTHORIZED,
        "FORBIDDEN": status.HTTP_403_FORBIDDEN,
        "NOT_FOUND": status.HTTP_404_NOT_FOUND,
    }
    return error_map.get(error_code, status.HTTP_500_INTERNAL_SERVER_ERROR)

